daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/transaction/PartialTransaction.scala

```diff
@@ -354,26 +364,67 @@ private[lf] case class PartialTransaction(
+          //
+          // Just like for modifying `keys` and `globalKeyInputs` we only consider
+          // by-key operations, i.e., lookup, exercise and fetch by key as well as creates
+          // and archives if the key has been brought into scope before.
+          //
+          // In the end, those checks mean that ledgers only have to look at inputs and outputs
+          // of the transaction and check for conflicts on that while speedy checks for internal
+          // conflicts.
+          //
+          // We have to consider the following cases for conflicts:
+          // 1. Create of a new local contract
+          //    1.1. KeyInactive in `keys`. This means we saw an archive so the create is valid.
+          //    1.2. KeyActive(_) in `keys`. This can either be local contract or a global contract. Both are an error.
+          //    1.3. No entry in `keys` and no entry in `globalKeyInputs`. This is valid. Note that the ledger here will then
+          //         have to check when committing that there is no active contract with this key before the transaction.
+          //    1.4. No entry in `keys` and `KeyInactive` in `globalKeyInputs`. This is valid. Ledgers need the same check
+          //         as for 1.3.
+          //    1.5. No entry in `keys` and `KeyActive(_)` in `globalKeyInputs`. This is an error. Note that the case where
+          //         the global contract has already been archived falls under 1.2.
+          // 2. Global key lookups
+          //    2.1. Conflicts with other global contracts cannot arise as we query a key at most once.
+          //    2.2. Conflicts with local contracts also cannot arise: A successful create will either
+          //         2.2.1: Set `globalKeyInputs` to `KeyInactive`.
+          //         2.2.2: Not modify `globalKeyInputs` if there already was an entry.
+          //         For both of those cases `globalKeyInputs` already had an entry which means
+          //         we would use that as a cached result and not query the ledger.
+          val (globalKeyInputs, conflict) = keys.get(ck) match {
+            // This is case 1.1.
+            case Some(KeyInactive) => (ptx.globalKeyInputs, KeyConflict.None)
+            // This is case 1.2.
+            case Some(KeyActive(_)) => (ptx.globalKeyInputs, KeyConflict.Duplicate)
+            case None =>
+              ptx.globalKeyInputs.get(ck) match {
+                // This is case 1.3.
+                case None => (ptx.globalKeyInputs.updated(ck, KeyInactive), KeyConflict.None)
+                // This is case 1.4.
+                case Some(KeyInactive) =>
+                  (ptx.globalKeyInputs.updated(ck, KeyInactive), KeyConflict.None)
+                // This is case 1.5
+                case Some(KeyActive(_)) =>
+                  (ptx.globalKeyInputs.updated(ck, KeyInactive), KeyConflict.Duplicate)
+              }
+
+          }
```

Comment:
I am wondering if it would not be simpler to split computation of `conflict` from the update of `globalKeyInputs` with something like that: 

```suggestion
          val conflict =  keys.get(ck).orElse(ptx.globalKeyInputs(ck) match {
             case Some(KeyActive(_)) => KeyConflict.Duplicate
             case Some(Inactive) | None => KeyConflict.None
          }
          val globalKeyInputs = keys.get(ck) match {
            case None => ptx.globalKeyInputs.updated(ck, KeyInactive)
            case Some(_) => ptx.globalKeyInputs
          }
```

Reply:
I’ve tried that first but it seemed more confusing and easier to get out of sync. Enumerating all the possible cases explicitly and doing so only once seems like an easy way to not mess it up.

Reply:
In particular, this way there is a very clear correspondence between the pattern matches and the doc comment which explains the details and the reasoning. At least to me that correspondence is pretty useful.

Reply:
Switched over to your suggestion (with some minor tweaks & fixes) which also fixed a bug. Thanks for the suggestion and for catching this!